cmd/argocd/commands/project.go

```diff
@@ -73,6 +93,288 @@ func addProjFlags(command *cobra.Command, opts *projectOpts) {
+func NewProjectRoleCreateTokenCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
+	var (
+		timeBeforeExpiry string
+	)
+	var command = &cobra.Command{
+		Use:   "create-token PROJECT TOKEN-NAME [--seconds seconds]",
+		Short: "Create a project token",
+		Run: func(c *cobra.Command, args []string) {
+			if len(args) != 2 {
+				c.HelpFunc()(c, args)
+				os.Exit(1)
+			}
+			projName := args[0]
+			roleName := args[1]
+			conn, projIf := argocdclient.NewClientOrDie(clientOpts).NewProjectClientOrDie()
+			defer util.Close(conn)
+			duration, err := timeutil.ParseDuration(timeBeforeExpiry)
+			errors.CheckError(err)
+			token, err := projIf.CreateToken(context.Background(), &project.ProjectTokenCreateRequest{Project: projName, Role: roleName, SecondsBeforeExpiry: int64(duration.Seconds())})
+			errors.CheckError(err)
+			fmt.Print(token.Token)
+		},
+	}
+	command.Flags().StringVarP(&timeBeforeExpiry, "timeBeforeExpiry", "s", "0s", "Time before the token will expire. (Default: No expiration)")
+
+	return command
+}
+
+// NewProjectRoleDeleteTokenCommand returns a new instance of an `argocd proj role delete-token` command
+func NewProjectRoleDeleteTokenCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
+	var command = &cobra.Command{
+		Use:   "delete-token PROJECT ROLE-NAME ISSUED_AT",
+		Short: "Delete a project token",
+		Run: func(c *cobra.Command, args []string) {
+			if len(args) != 3 {
+				c.HelpFunc()(c, args)
+				os.Exit(1)
+			}
+			projName := args[0]
+			roleName := args[1]
+			issuedAt, err := strconv.ParseInt(args[2], 10, 64)
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			conn, projIf := argocdclient.NewClientOrDie(clientOpts).NewProjectClientOrDie()
+			defer util.Close(conn)
+
+			_, err = projIf.DeleteToken(context.Background(), &project.ProjectTokenDeleteRequest{Project: projName, Role: roleName, IssuedAt: issuedAt})
+			errors.CheckError(err)
+		},
+	}
+	return command
+}
+
+// NewProjectRoleListCommand returns a new instance of an `argocd proj roles list` command
+func NewProjectRoleListCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
+	var command = &cobra.Command{
+		Use:   "list PROJECT",
+		Short: "List all the roles in a project",
+		Run: func(c *cobra.Command, args []string) {
+			if len(args) != 1 {
+				c.HelpFunc()(c, args)
+				os.Exit(1)
+			}
+			projName := args[0]
+			conn, projIf := argocdclient.NewClientOrDie(clientOpts).NewProjectClientOrDie()
+			defer util.Close(conn)
+
+			project, err := projIf.Get(context.Background(), &project.ProjectQuery{Name: projName})
+			errors.CheckError(err)
+			w := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
+			fmt.Fprintf(w, "ROLE-NAME\tISSUED-AT\tEXPIRES-AT\tPOLICIES\n")
+			for _, role := range project.Spec.Roles {
+				fmt.Fprintf(w, "%s\n", role.Name)
+				if role.JWTTokens != nil {
+					for _, token := range role.JWTTokens {
+						fmt.Fprintf(w, "%s\t%d\t%d\n", role.Name, token.IssuedAt, token.ExpiresAt)
```

Comment:
If a token does not expire, I think this is going to print epoch zero (1970). Instead we should print `<none>`

Reply:
Actually I see that it will print `0` since this is '%d'. Instead can we print the expiration time in a human readable format. e.g. 2018-08-15T12:13:14?

Reply:
Addressed this issue by breaking the list role command into a list role and a get role command.  The list role command would list all the roles and their descriptions.  The get role command would list the name, description, policies, and JWT tokens of a role in a project.  With the JWT token, the cli lists the issue-at value as a ID to used in the delete-token and two human readable tokens for issued-at and expires-at.

Reply:
Example output:
```
./dist/argocd proj role list guestbook
ROLE-NAME                    Description
test                         desc
asdfghjklasdfghjklasdfghjkl  desc1
```
```
./dist/argocd proj role get guestbook test
Role Name: test
Description:desc
Policies:
p, proj:guestbook:test, applications, get, guestbook/guestbook-test, allow
p, proj:guestbook:test, applications, get, guestbook/*, allow
Jwt Tokens:
ID          ISSUED-AT                                  EXPIRES-AT
1534270415  Tue Aug 14 11:13:35 -0700 (4 hours ago)    <none>
1534286505  2018-08-14T15:41:45-07:00 (16 minutes ago)  2018-08-24T15:41:45-07:00 (1 week from now)
1534286515  2018-08-14T15:41:55-07:00 (16 minutes ago)  2018-09-03T15:41:55-07:00 (2 weeks from now)
1534286520  2018-08-14T15:42:00-07:00 (16 minutes ago)  2018-09-04T15:42:00-07:00 (2 weeks from now)
1534286527  2018-08-14T15:42:07-07:00 (16 minutes ago)  2018-09-05T15:42:07-07:00 (3 weeks from now)
1534286539  2018-08-14T15:42:19-07:00 (15 minutes ago)  2018-10-13T15:42:19-07:00 (1 month from now)
1534286552  2018-08-14T15:42:32-07:00 (15 minutes ago)  2018-08-14T15:43:02-07:00 (15 minutes ago)
1534286554  2018-08-14T15:42:34-07:00 (15 minutes ago)  2018-08-14T15:43:04-07:00 (15 minutes ago)
1534286558  2018-08-14T15:42:38-07:00 (15 minutes ago)  2018-08-14T15:43:08-07:00 (15 minutes ago)```